blog/2023/06/PSA-dont-use-threadid.md

```diff
@@ -0,0 +1,174 @@
++++ mintoclevel = 2 maxtoclevel = 3 title = "PSA: Stop using `states[threadid()]`" authors = "Mason Protter, Valentin Churavy, Ian Butterworth, ..." published = "XX June 2023" rss_pubdate = Date(2023, 06, XX) rss = """PSA: Stop using `states[threadid()]`""" +++
+
+# PSA: Stop using `states[threadid()]`
+Alt titles:
+- PSA: Multithreading with `states[threadid()]` is unsafe
+- PSA: Multithreading with `buffers[threadid()]` is unsafe
+- PSA: Don't assume `threadid()` is stable within a task
+
+__
+
+
+Partially due to the evolving history of our parallel and concurrent interfaces[^historynote], some Julia programmers have been writing *incorrect* parallel code that contains the possibility of race conditions which can give wrong results. This pattern has even been erroneously recommended in [previous official julia blogposts](https://github.com/JuliaLang/www.julialang.org/blob/main/blog/2019/07/multithreading.md#thread-local-state). It is important for the stability and correctness of the ecosystem that these usages are identified and fixed.
+
+The general template that this incorrect code follows is something like the following:
+
+```julia
+using Base.Threads: nthreads, @threads, threadid
+
+states = [some_initial_value for _ in 1:nthreads()]
+@threads for x in some_data
+    tid = threadid()
+    old_val = states[tid]
+    new_val = some_operator(old_val, f(x))
+    states[tid] = new_val
+end
+do_something(states)
+```
+
+The above code is **incorrect** because the tasks spawned by `@threads` are  allowed to yield to other tasks during their execution. This means that between reading `old_val` and storing `new_val` in the storage, the task could be paused and a new task running on the same thread with the same `threadid()` could concurrently write to `states[tid]`, causing a race condition and thus work being lost.
```

Comment:
Would it be a good idea to be more precise here, mentioning how adding an innocous looking debug printout in `f` or `some_operator` can induce the race conditions here, due to adding a yieldpoint? Or is it better not to mention what kind of stuff can make a task yield, since we're a cooperative runtime and not preemptive?

Reply:
Maybe that's a good use for a footnote or something?

Reply:
For developers who may be less familiar with race conditions, would it be easier to delay explaining what can go wrong until the `f(i) = (sleep(0.001); i)` example below? I'm imagining you could do something like this:
```julia
    tid = Threads.threadid()   # each task gets `tid = 1`
    old_var = state[tid]       # each task reads the current value, which for all is 0 (!) because... 
    new_var = old_var + f(i)   # ...the `sleep` in `f` causes all tasks to pause *simultaneously* here (all loop iterations start, but do not yet finish)
    state[tid] = new_var       # after being released from the `sleep`, each task sets `state[1]` to 1
```
My hope is that the comments walk readers through the logic more clearly than the short description above.

Reply:
hm yeah that probably is more clear than the prose. What would you think about having the comments *and* the prose?

Reply:
In addition to your comments Tim, @Seelengrab  I've added a new footnote trying to explain that while some code is currently okay because it happens to not yield, this is not a viable path forward and it could break at any time.